### ISSUE TITLE
Prevent testnet to mainnet swaps

### ISSUE THREAD

**Issue by D4nte** [comit-network/comit-rs#376]
Node should reject *incompatible* network request (ie, testnet to mainnet).

Probably need to introduce a "type" of networks (mainnet, testnet).

Note:
Future feature could be to make it configurable. Not part of this ticket.

Also do #710 
Blocked by #644

**Comment by thomaseizinger** [comment]
Another way we discussed at some point is to define a `trait` that all `Network`s (or `Ledger`s) have to implement to determine the compatibility.

For example:

```rust
pub trait Network<O: Network> {
   fn is_safe_to_swap_with(&self, other: O) -> bool
}

impl Network for bitcoin::Network<ethereum::Network> {
   fn is_safe_to_swap_with(&self, other: ethereum::Network) -> bool {
       match self {
           bitcoin::Network::Mainnet => other == ethereum::Network::Mainnet,
           // You can imagine the rest
       }
   }
}
```

This would allow us to generically check if swapping from one Ledger to another Ledger is "safe".

**Comment by D4nte** [comment]
Please note that the various Ethereum testnets (Ropsten, Rinkeby) should be considered compatible with Bitcoin testnet

**Comment by bonomat** [comment]
I'm against preventing testnet to mainnet swaps automatically from within comit-rs. In fact, this could be even a desired behavior: e.g. even testnet bitcoins have some kind of value. 

I'd rather see this as a feature of the UI: It should highlight in red that a requested  swap (or a swap you are about to request) is between a mainnet and a non-mainnet. 
This would simplify things as we don't need to make assumptions between different testnet compatibilities (Ethereum's Ropsten, Rinkbey, etc is compatible with Bitcoin's Testnet, etc. ).

**Comment by D4nte** [comment]
"even testnet bitcoins have some kind of value." How much do you want to buy my testnet bitcoins?

**Comment by thomaseizinger** [comment]
I think it should be prevented automatically and you can opt-in to it with a commandline flag or something that says `--allow-cross-network-swaps`.

**Comment by D4nte** [comment]
I like @thomaseizinger 's idea

**Comment by bonomat** [comment]
> "even testnet bitcoins have some kind of value." How much do you want to buy my testnet bitcoins?

I don't say it is much, but since most testnets do PoW they have some value. 

> I think it should be prevented automatically and you can opt-in to it with a commandline flag or something that says --allow-cross-network-swaps.

A few thoughts on that: 
* I think a comit node (comit-rs) should not be opinionated about the value of an asset. 
* The decision of accepting/declining swap requests should be application layer logic as there are too many combinations we would need to consider. Ethereum  only has already 3 different testnets (Ropsten, Rinkeby and Kovan). 
* `--allow-cross-network-swaps`, is not clear enough for me, e.g. does this mean Bitcoin's testnet to Ethereum Ropsten is valid by default, but Bitcoin's testnet and Ethereum Rinkeby not (or as well). Rinkeby is a Proof-of-Authority network so actually different to Ropsten

One use case from the past: Cryptokitties started on Ropsten first and if you were really good you could win some Ether (+ you had a head start on their mainnet launch), hence, people sold already their Ropsten kitties ;)

**Comment by thomaseizinger** [comment]
> --allow-cross-network-swaps, is not clear enough for me, e.g. does this mean Bitcoin's testnet to Ethereum Ropsten is valid by default, but Bitcoin's testnet and Ethereum Rinkeby not (or as well). Rinkeby is a Proof-of-Authority network so actually different to Ropsten

The naming can obviously be revised it is more about the idea of having such a flag. You probably just want to differentiate between mainnet and not-mainnet.

> The decision of accepting/declining swap requests should be application layer logic as there are too many combinations we would need to consider. Ethereum only has already 3 different testnets (Ropsten, Rinkeby and Kovan).

This means, _each_ application trying to implement the same security measures would have to implement all of these _many combinations_. I'd prefer the comit node to do this check for you and make the clients more stupid.

> I think a comit node (comit-rs) should not be opinionated about the value of an asset.

My thinking process on whether the software should have an opinion on something is usually as follows:

1. If you would have to acquire more information from the user to make the decision, you probably want the decision to be made outside.
2. If there is not an obvious choice to be made, you probably want the decision to be made by the user instead of trying to be "clever".

In our case, we already know about the concept of networks, so point 1 is not a problem. Not wanting to swap between mainnet and testnet is very likely what the user wants or put the other way round: Cases where the user would want that are pretty rare.

Based on that, I think it is fine for the software to have an opinion about it.

**Comment by bonomat** [comment]
Revisit this once comit/btsieve can connect to multiple networks of the same ledger.
